## python library 

# NumPy 
# a library for scientific computing in python 
# come pre-installed with anaconda enviroment but installed seperatelty ie for 
# provides support for large multi deimensional arrays and matrices 
# contains high level mathematical 
# created by travis oliphant in 2005

# can be imported in python code:
#         import numpy as np
#  


# teacher - hard to visuallise all the code in ur head so google 

In [2]:
# NumPy Fundamentals 

import numpy as np


In [3]:
## 1 dimensional array 

m1darray = np.array([1,2,3])
print(m1darray)


[1 2 3]


In [5]:
## 2 dimensional array 
my2darray = np.array([[1,2,3,], [4,5,6]])
print(my2darray)

#here with two array we need double square brackets 
# 2 dimenion is list of lists 

[[1 2 3]
 [4 5 6]]


In [8]:
m1darray.dtype


dtype('int64')

In [31]:
my2darray.dtype


dtype('int64')

In [30]:
## specifying the data type
# ie into a float
my2darrayfloat = np.array([[1,2,3,],[4,5,6]],dtype='float16')
print(my2darrayfloat)

#dtype = 'float16' why cant it be 15 or 12


#chat gpt
#float16: 16 bits (half precision)
#float32: 32 bits (single precision)
#float64: 64 bits (double precision)

[[1. 2. 3.]
 [4. 5. 6.]]


In [25]:
my2darrayfloat.dtype


dtype('float32')

In [ ]:
my2darrayfloat = np.array([[1,2,3,],[4,5,6]],dtype='float16')
print(my2darrayfloat)

In [36]:
# if your array has a conbination of integer float and string 
# square bracket is for list
# we using both normal bracket and square ones as its contains multiple data types 


charArray1 = np.array([(1, 'a', 3), (4, 5, 'zzz')])

charArray2 = np.array([(1, 'a', 3), (4, 5, 'zzz')], dtype = 'U11')
# second line to assign what we want the dtype to be although i dont understand why
#python prioritise strings over integers

# look back

In [35]:
charArray1.dtype

#also u have to run charArray1 before it can find the dtype
#dtype('<U21') u means its a unicode
# U11
#whats a unicode
# so it represent the max number of values,string length u should have in that array 


dtype('<U21')

In [37]:
charArray2.dtype

dtype('<U11')

In [39]:
# array of initialization 

number_rows = 2
number_columns = 3 

#array of 0 
arrays0 = np.zeros([number_rows, number_columns])
arrays0 = np.zeros([2, 3])
arrays0 = np.zeros([number_rows, number_columns], dtype=np.int32)

print(arrays0)

#all prints out the same thing 


#but why does it give just 0
# so after testing it out u can change the word zero to ones or zeros then it will change

[[0. 0. 0.]
 [0. 0. 0.]]


In [55]:
arrays0 = np.zeros([number_rows, number_columns], dtype=np.float32)
# u can change it to a float dtype instead of int

In [47]:
# array of ones 
arrays1 = np.ones([number_rows, number_columns])
arrays2 = 5 * np.ones([number_rows, number_columns])
print(arrays1)
print(arrays2)
#
#

[[1. 1. 1.]
 [1. 1. 1.]]
[[5. 5. 5.]
 [5. 5. 5.]]


In [49]:
## arrays of 7 
array7 = np.full([number_rows,number_columns], 7)
print(array7)

#so with this the changes are full function and the 7 at the end of the bracket 
#number on the list can be changed by change the number the end. 
#full will fill with ur desired number 
#rather than multiplyin it at the beginning of the code

[[8 8 8]
 [8 8 8]]


# CHECK THIS BELOW

In [ ]:
#list vs toggle


#
# toggle are more faster? effiencet at data?

In [53]:
# WE CAN also set parameters 

start = 10
end = 27
step = 3
evenstep = np.arange(start, end, step)
#new one is arange 
print(evenstep)

# cause end is 27, the last numeber willl also be ignored

[10 13 16 19 22 25]


In [54]:
# genrating equall spaced number 
number_sample = 4
evensample = np.linspace(start, end, number_sample)
# linspace generatte number with 4 sample
print(evensample)

#so in this from 10-27 its been split into 4 different sample equally so add up every 5.66666667

[10.         15.66666667 21.33333333 27.        ]


In [56]:
#indetity matrix between 0 and 1
identity = np.eye(3)
print(identity)

#whar is np.eye a 2x2 identity matrix

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [68]:
# random matrix
# will generate floating numbers between 0 and 1, 1 exclusive so will not be included 
random01 = np.random.random((2,3)) # obvs 2,3 number of rows, columns 
print(random01)

#everytime u run it will be different between 0 and 1 

[[0.00600554 0.96565906 0.92816693]
 [0.95591529 0.9941833  0.06548404]]


# CHECK QUESTION BELOW DUONG

In [83]:
# random.randint
# this time with specifc we want it between 00 and 10

randomintr = np.random.randint(0,10,(2,3)) # comma is important  # ask chat why
print(randomintr)
# SO u will find 0-9 but not 10 in output
# last number not included

[[2 2 4]
 [0 1 8]]


# QUESTION DUONG BELOW

In [91]:
#indexing, inpecting and slicing 


a = np.array(([5,6,3],[2,5,1]))
print(a)
print('a has dimension:', a.ndim) # THIS property on the side called what the array has
print('a has dimension:', a.shape)
print('a has dimension:', a.size)
print('a has dimension:', a.dtype)


#why is mine always by default dtype 64 and not 32 or otherss 

[[5 6 3]
 [2 5 1]]
a has dimension: 2
a has dimension: (2, 3)
a has dimension: 6
a has dimension: int64


In [ ]:
#

In [92]:
#indexing 
a[1,2]

1

In [93]:
a[0:2,1]

#: double colon means 0to2 but excluding 2

array([6, 5])

In [94]:
a[0,:3]

array([5, 6, 3])

# indexing was confusing look over above

In [95]:
#diagonal 
np.diagonal(a)

#a based on array above

array([5, 5])

In [98]:
#diagonal using another array b

b = np.array([[5,6,3],[2,5,1],[4,4,4,]]) #dont forgrt brackets 
print(b)

[[5 6 3]
 [2 5 1]
 [4 4 4]]


In [99]:
np.diag(b)# shows diagonal numbers

array([5, 5, 4])

In [ ]:
# another one 


In [ ]:
#task 1 
arr = [0,2,4,5,3,7,8,9,5,1]

num = np.full(arr,arr)
if arr % 2 == 1:
    print(num)

In [ ]:
arr2 = np.array[(0,2,4,5,3,7,8,9,5,1)]

arr2[0,:3]


# LOOL I NEED MAD HELP ON THIS ONE

In [121]:
# task odd numbers , replace wiht negative valunes tast 1 answer from teacher

num_array2 = np.array([0,2,4,5,3,7,8,9,5,1])# girll stop mistaking ur fing brackets square on the inside!!!!!

## wrtie a function to calculate the odd nubers 
odd_num = np.where(num_array2 % 2 == 1)

#replace odd number with negative value 
num_array2[odd_num] *=-1 

print(num_array2)


#where fuction is ??

[ 0  2  4 -5 -3 -7  8 -9 -5 -1]


In [127]:
a1 = np.array([[1,2,4],[7,1,3]])
a2 = np.array([[4,6,9],[2,5,1]])

#addtion
addition = a1 + a2 
addition1 = np.add(a1,a2)
print(addition)
print(addition1)

[[ 5  8 13]
 [ 9  6  4]]
[[ 5  8 13]
 [ 9  6  4]]


In [128]:
# following on
#subtraction 

sub1 = a1 -a2
sub2 = np.subtract(a1,a2)
print(sub1)
print(sub2)

[[-3 -4 -5]
 [ 5 -4  2]]
[[-3 -4 -5]
 [ 5 -4  2]]


In [129]:
## following on 
# multplication 
multi = a1 * a2
multi2 = np.multiply(a1,a2)
print(multi)
print(multi2)

[[ 4 12 36]
 [14  5  3]]
[[ 4 12 36]
 [14  5  3]]


In [130]:
## following on 
#division 

divi = a1 / a2
divi1 = np.divide(a1,a2)
print(divi)
print(divi1)

[[0.25       0.33333333 0.44444444]
 [3.5        0.2        3.        ]]
[[0.25       0.33333333 0.44444444]
 [3.5        0.2        3.        ]]


In [132]:
## following on 
# dot product = whats dot product 
# well it is 

a3 = np.array([[4,6,],[2,5],[9,1]])
dotp = a1.dot(a3)
print(dotp)




[[44 20]
 [57 50]]


In [136]:
## following 
# other calculation 
# cos, sin, log, sqroot , exp 

exp = np.exp(a)
sqroot = np.sqrt(a)
sin = np.sin(a)
cos = np.cos(a)
log = np.log(a)

print(exp)
print(sqroot)
print(sin)
print(cos)
print(log)

[[148.4131591  403.42879349  20.08553692]
 [  7.3890561  148.4131591    2.71828183]]
[[2.23606798 2.44948974 1.73205081]
 [1.41421356 2.23606798 1.        ]]
[[-0.95892427 -0.2794155   0.14112001]
 [ 0.90929743 -0.95892427  0.84147098]]
[[ 0.28366219  0.96017029 -0.9899925 ]
 [-0.41614684  0.28366219  0.54030231]]
[[1.60943791 1.79175947 1.09861229]
 [0.69314718 1.60943791 0.        ]]


In [150]:

a5 = np.array([[1,2,5],[7,1,3]])
std = np.std(a5)
sum1 = np.sum(a5)
sum2 = a5.sum(axis=1)
sum3 = np.sum(a5, axis=1) #same thing adds 1st row, and the rest in one
maximum = a5.max()
maximum2 = a5.max(axis=0)

#average 
avg = np.mean(a5) 
avg2 = np.mean(a5, axis=0)

print(std)
print(sum1)
print(sum2)
print(sum3)
print(maximum)
print(maximum2)

print(avg)
print(avg2)

2.1921577396609844
19
[ 8 11]
[ 8 11]
7
[7 2 5]
3.1666666666666665
[4.  1.5 4. ]


In [154]:
## triangle 
A = np.array([0,0])
B = np.array([0,3])
C = np.array([1.5,0])

from numpy import linalg as la 
AB = la.norm(A-B)
AC = la.norm(A-C)
BC = la.norm(B-C)

#now u can print output 
print(AB)
print(AC)
print(BC)



3.0
1.5
3.3541019662496847
1.7320508075688772
1.224744871391589
1.8314207507423532
2.956795678960466


In [ ]:
# i think i did pythagoous thoerem wrong 

sqroot = np.sqrt(AB)
sqroot1 = np.sqrt(AC)
sqroot2 = np.sqrt(BC)

ptheorm = sqroot + sqroot1

print(sqroot)
print(sqroot1)
print(sqroot2)

print(ptheorm)

In [156]:
#new array

a7 = np.array([[1,2,5],[7,1,3]])

## transporting

atr1 = np.transpose(a7)
atr2 = a7.T

print(a7)
print(atr2)


[[1 2 5]
 [7 1 3]]
[[1 7]
 [2 1]
 [5 3]]


In [159]:
## reshape 

reshape1 = np.reshape(a7, (3,2))
row = a7.reshape(6)
print(reshape1)
print(row)

[[1 2]
 [5 7]
 [1 3]]
[1 2 5 7 1 3]


In [161]:

## brain got fried in 5:50 mins in video 
a8 = np.array([[1,2,5],[7,1,3]])
a9 = np.array([[7,8],[4,1]])
a10 = np.array([[9],[9]])

abc = np.hstack([a8,a9,a10]) #hstack horizontal order
print(abc)

[[1 2 5 7 8 9]
 [7 1 3 4 1 9]]


In [164]:
#stacking but vstack
array1 = np.array([[1,2,3],[4,5,6]])
array2 = np.array([[7,8,9],[10,11,12]])
array3 = np.array([[13,14,15],[16,17,18]])

asd = np.vstack([array1,array2,array3]) #vstack vertical order
print(asd)

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]
 [13 14 15]
 [16 17 18]]


In [167]:
## concatenate 
ab = np.concatenate((array1,array2,array3), axis=1) #1 axis is like horizontal
print(ab)

[[ 1  2  3  7  8  9 13 14 15]
 [ 4  5  6 10 11 12 16 17 18]]


In [168]:
## concatenate 
ab1 = np.concatenate((array1,array2,array3), axis=0) #0 axis is like veritcal
print(ab1)

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]
 [13 14 15]
 [16 17 18]]


In [169]:
## save a single array to a file 
np.save('array a', array1)

#so this binary file will be saved instead of .py or .iynb its .npy 
# you wouldnt be able to open it w/o create some code 


In [170]:

#this savez you can save numerous array in one file
np.savez('array_ab', arr1 = array1, arr2 = array2)

In [171]:
#savetxt will

np.savetxt('array_a.txt', array1)
 #this one will show a the array 

In [175]:
# delimiter
## %d = integer format
## %f === float number
## %s == string number
## %.2f --- float number to the nearest 2 decimal place 
np.savetxt('array_a4.txt', array1, delimiter=',', fmt='%.2f')

In [178]:
# save to a csv file 
np.savetxt('array.csv', array2,delimiter= ',' , fmt='%d')

#also what is a csv file 

In [179]:
# how to open the binary file of just a array 
# single array and doube array 

## to load .npy file 

le = np.load('array a.npy')
print(le)

 


[[1 2 3]
 [4 5 6]]


In [183]:
## load the npz file 

lez = np.load('array_ab.npz')
print(lez)
print(lez['arr1'])
print(lez['arr2'])

NpzFile 'array_ab.npz' with keys: arr1, arr2
[[1 2 3]
 [4 5 6]]
[[ 7  8  9]
 [10 11 12]]


In [190]:
## load text file 
read_text = np.loadtxt('array_a4.txt', delimiter=',')
print(read_text)

#this ones needs a delimiter? in order to read it 

[[1. 2. 3.]
 [4. 5. 6.]]


In [195]:
## load a csv file 


read_textcsv = np.loadtxt('array.csv', delimiter=',')
print(read_textcsv)

[[ 7.  8.  9.]
 [10. 11. 12.]]


In [199]:
# task 2 

new_array = np.array([[1.89, 2.23, 3.35], [5.30, 7.75,5.65], [1.89, 3.01, 4.56]])

## round of the array to the nearest 1 decimal place
new_array = np.round(new_array, decimals=1)

#add column header 
column_header = ['col1', 'col2', 'col3']

# add row header 

row_header = ['----'], ['row1'], ['row2'], ['row3']

## stacking rows and columns header to the arrays 
array_with_header = np.vstack([column_header, new_array])

final_array = np.hstack([row_header,array_with_header])

# save the array to a csv file 
np.savetxt('arrayfile.csv', final_array, delimiter=',', fmt='%s')







# pip3 install numpy for mac - all done ! 
#      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.2/61.2 kB 1.1 MB/s eta 0:00:00
# Downloading numpy-1.26.3-cp312-cp312-macosx_10_9_x86_64.whl (20.3 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.3/20.3 MB 6.0 MB/s eta 0:00:00
# Installing collected packages: numpy
# Successfully installed numpy-1.26.3

# [notice] A new release of pip is available: 23.2.1 -> 23.3.2
# [notice] To update, run: python3.12 -m pip install --upgrade pip (should i update it)
# code to update was pip3 install --upgrade pip DONE!!
